dht11_reader.c

```diff
@@ -0,0 +1,403 @@
+// file input.c
+#include <gpiod.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <string.h>
+#include <limits.h>
+#include "getopt.h"
+
+static const char *prog_name = "dht11_reader";
+
+/* timespec_diff is a method for taking the 
+ * difference between two timespecs.  
+ * Note that stop >= start, 
+ * otherwise the resulting timespec will not be valid.
+ */
+struct timespec timespec_diff(
+  struct timespec *start, 
+  struct timespec *stop
+)
+{
+    struct timespec result;
+    if ((stop->tv_nsec - start->tv_nsec) < 0) {
+        result.tv_sec = stop->tv_sec - start->tv_sec - 1;
+        result.tv_nsec = stop->tv_nsec - start->tv_nsec + 1000000000;
+    } else {
+        result.tv_sec = stop->tv_sec - start->tv_sec;
+        result.tv_nsec = stop->tv_nsec - start->tv_nsec;
+    }
+
+    return result;
+}
+
+struct config {
+  unsigned int line_offset;
+  unsigned int hold_period_us;
+  bool verbose;
+  unsigned int chip_number;
+};
+
+// The following is taken from 
+// https://git.kernel.org/pub/scm/libs/libgpiod/libgpiod.git/tree/tools/tools-common.c
+void print_period_help(void)
+{
+        printf("\nPeriods:\n");
+        printf("    Periods are taken as milliseconds unless units are specified. e.g. 10us.\n");
+        printf("    Supported units are 's', 'ms', and 'us'.\n");
+}
+
+static void print_help(void)
+{
+        printf("Usage: %s [OPTIONS] \n", prog_name);
+        printf("\n");
+        printf("Read DHT11 sensor.\n");
+        printf("\n");
+        printf("Options:\n");
+        printf("  -l, --line <offset>           line offset for the DHT11 signal              default: 17\n");
+        printf("  -p, --hold-period <period>    time period to hold low to initiate sensor    default: 180\n");
+        printf("  -c, --chip <chip>             restrict scope to a particular chip           default: 0\n");
```

Comment:
Make sure the explanation makes it clear this is the chip number, not a full device path or some other string.